include/net/virtual.h

```diff
@@ -0,0 +1,335 @@
+		struct in6_addr peer6addr;
+		int mtu;
+	};
+};
+
+#if defined(CONFIG_NET_L2_VIRTUAL)
+#define VIRTUAL_MAX_NAME_LEN CONFIG_NET_L2_VIRTUAL_MAX_NAME_LEN
+#else
+#define VIRTUAL_MAX_NAME_LEN 0
+#endif
+/** @endcond */
+
+struct virtual_interface_api {
+	/**
+	 * The net_if_api must be placed in first position in this
+	 * struct so that we are compatible with network interface API.
+	 */
+	struct net_if_api iface_api;
+
+	/** Get the virtual interface capabilities */
+	enum virtual_interface_caps (*get_capabilities)(struct net_if *iface);
+
+	/** Start the device */
+	int (*start)(const struct device *dev);
+
+	/** Stop the device */
+	int (*stop)(const struct device *dev);
+
+	/** Send a network packet */
+	int (*send)(struct net_if *iface, struct net_pkt *pkt);
+
+	/** Receive a network packet */
+	enum net_verdict (*recv)(struct net_if *iface, struct net_pkt *pkt);
+
+	/** Check if this received network packet is for this interface.
+	 * The callback returns NET_CONTINUE if this interface will accept the
+	 * packet and pass it upper layers, NET_DROP if the packet is to be
+	 * dropped and NET_OK to pass it to next interface.
+	 */
+	enum net_verdict (*input)(struct net_if *input_iface,
+				  struct net_if *iface,
+				  struct net_addr *remote_addr,
+				  struct net_pkt *pkt);
+
+	/** Pass the attachment information to virtual interface */
+	int (*attach)(struct net_if *virtual_iface, struct net_if *iface);
+
+	/** Set specific L2 configuration */
+	int (*set_config)(struct net_if *iface,
+			  enum virtual_interface_config_type type,
+			  const struct virtual_interface_config *config);
+
+	/** Get specific L2 configuration */
+	int (*get_config)(struct net_if *iface,
+			  enum virtual_interface_config_type type,
+			  struct virtual_interface_config *config);
+};
+
+/* Make sure that the network interface API is properly setup inside
+ * Virtual API struct (it is the first one).
+ */
+BUILD_ASSERT(offsetof(struct virtual_interface_api, iface_api) == 0);
+
+/** Virtual L2 context that is needed to binding to the real network interface
+ */
+struct virtual_interface_context {
+/** @cond INTERNAL_HIDDEN */
+	/* Keep track of contexts */
+	sys_snode_t node;
+
+	/* My virtual network interface */
+	struct net_if *virtual_iface;
+/** @endcond */
+
+	/**
+	 * Other network interface this virtual network interface is
+	 * attached to. These values can be chained so virtual network
+	 * interfaces can run on top of other virtual interfaces.
+	 */
+	struct net_if *iface;
+
+	struct {
+		/** Carrier ON/OFF handler worker. This is used to create
+		 * network interface UP/DOWN event when ethernet L2 driver
+		 * notices carrier ON/OFF situation. We must not create another
+		 * network management event from inside management handler thus
+		 * we use worker thread to trigger the UP/DOWN event.
+		 */
+		struct k_work work;
+
+		/** Network interface that is detecting carrier ON/OFF event.
+		 */
+		struct net_if *iface;
+	} carrier_mgmt;
```

Comment:
I think I will remove the carrier handling as that is handled by the real network interface.